### PR TITLE
Allow callers to track add/remove listeners

### PR DESCRIPTION
**Pull request by ddumont**
Something like this may help callers clean up listeners #32

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -185,43 +194,51 @@ function renderMedia (file, getElem, opts, cb) {
...
-      elem.addEventListener('loadstart', onLoadStart)
-      elem.addEventListener('canplay', onCanPlay)
+      addListener(elem, 'error', fallbackToBlobURL)
+      addListener(elem, 'loadstart', onLoadStart)
+      addListener(elem, 'canplay', onCanPlay)
 
-      var wrapper = new MediaElementWrapper(elem)
-      var writable = wrapper.createWriteStream(getCodec(file.name))
-      file.createReadStream().pipe(writable)
+      try {
```
Comment (feross) [inline index.js:210]: Why did you introduce a `try` here?

File: index.js [hunk 2]
```diff
@@ -185,43 +194,51 @@ function renderMedia (file, getElem, opts, cb) {
...
-      elem.addEventListener('canplay', onCanPlay)
-      getBlobURL(file, function (err, url) {
-        if (err) return fatalError(err)
-        elem.src = url
-        if (currentTime) elem.currentTime = currentTime
-      })
+      addListener(elem, 'error', fatalError)
+      addListener(elem, 'loadstart', onLoadStart)
+      addListener(elem, 'canplay', onCanPlay)
+      try {
```
Comment (feross) [inline index.js:227]: This is not how to handle errors. We already pass `err` into the callback.

File: index.js [hunk 3]
```diff
@@ -169,12 +169,21 @@ function renderMedia (file, getElem, opts, cb) {
...
+    opts.onRemoveListener(elem, event, listener)
+  }
+
   function renderMediaSource () {
     var tagName = MEDIASOURCE_VIDEO_EXTS.indexOf(extname) >= 0 ? 'video' : 'audio'
 
     if (MediaSource) {
       if (VIDEOSTREAM_EXTS.indexOf(extname) >= 0) {
-        useVideostream()
+        return useVideostream()
```
Comment (feross) [inline index.js:186]: It's not ideal to return the `Videostream` instance in this case but to return nothing in the other cases. The user of this API will be very confused.

File: index.js [hunk 4]
```diff
@@ -185,43 +194,51 @@ function renderMedia (file, getElem, opts, cb) {
...
       debug('Use `videostream` package for ' + file.name)
       prepareElem()
-      elem.addEventListener('error', fallbackToMediaSource)
-      elem.addEventListener('loadstart', onLoadStart)
-      elem.addEventListener('canplay', onCanPlay)
-      videostream(file, elem)
+      addListener(elem, 'error', fallbackToMediaSource)
+      addListener(elem, 'loadstart', onLoadStart)
+      addListener(elem, 'canplay', onCanPlay)
+      return new Videostream(file, elem)
```
Comment (feross) [inline index.js:200]: Why are you returning the `Videostream` instance?

**Review by feross — CHANGES_REQUESTED** [review]
You're missing a `return renderMedia` in the `append()` function.

---

But more generally, I think this approach is exposing too much of the implementation details to the user. We should just return a single function that, when called, removes and destroys all resources.

**Comment by feross** [comment]
@ddumont Are you interested in addressing the review feedback in this PR and helping get it over the finish line?

**Comment by ddumont** [comment]
I'd like to close this pr, I've since moved away from render-media.

**Pull request closed without merging**